Update: Known array lengths

Harris-Miller opened the pull request:
As pointed out in https://github.com/ramda/ramda/issues/3364, V8 cares if you instantiate an Array with a known length is far faster to fill than to start with an empty array

This MR updates the following areas of `ramda` to do exactly that

* _clone :: https://jsperf.app/jopaqu - seeing a ~10% gain
* intersperse :: https://jsperf.app/dibehu - ~40%
* mapAccum :: https://jsperf.app/gozalu - 2% (small)
* mapAccumRight, same as ^^
* range :: https://jsperf.app/yaxihu - ~45%
* scan :: https://jsperf.app/tokefo - ~50%
* times :: https://jsperf.app/matoqu - ~50% gain!
* xprod :: https://jsperf.app/cezeko - ~75%
* zip :: https://jsperf.app/foxowi - seeing a ~15% gain
* zipWith :: https://jsperf.app/yasibi - ~ 20%

Berndy commented:
These seem like some useful improvements, ramda is in dire need of performance. Maybe sync with master and declare it ready for review?

kedashoe commented on source/intersperse.js:
```diff
@@ -19,15 +19,17 @@ import _curry2 from './internal/_curry2.js';
-  var idx = 0;
+var intersperse = _curry2(_checkForMethod('intersperse', function _intersperse(separator, list) {
   var length = list.length;
+  var out = Array(length * 2 - 1);
```
This will throw an error if intersperse is passed an empty array. We can just bail early and return empty array in that case.

Harris-Miller replied:
Updated with a bail out when `length === 0`

kedashoe commented on source/xprod.js:
```diff
@@ -24,11 +24,11 @@ var xprod = _curry2(function xprod(a, b) { // = xprodWith(prepend); (takes about
   var j;
   var jlen = b.length;
-  var result = [];
+  var result = Array(ilen + jlen);
```
This should be multiplication instead of addition yes? `Array(ilen * jlen)`. I suspect the huge performance gains for xprod have something to do with this..

Harris-Miller replied:
Updated

kedashoe commented on source/xprod.js:
```diff
@@ -24,11 +24,11 @@ var xprod = _curry2(function xprod(a, b) { // = xprodWith(prepend); (takes about
     j = 0;
     while (j < jlen) {
-      result[result.length] = [a[idx], b[j]];
+      result[idx + j] = [a[idx], b[j]];
```
And here in the inner loop, we are adding `jlen` elements each time through, so I believe we have

`result[(idx * jlen) + j] = [a[idx], b[j]];`

It is also a bit confusing to read with `idx` and `j` as the index variables. Can we update this function to use just plain old `i` to match the `j` (I know you didn't name these originally)

Harris-Miller replied:
Updated

kedashoe reviewed (changes requested):
Thanks for the PR @Harris-Miller , sorry I didn't review this initially!

kedashoe commented:
> These seem like some useful improvements, ramda is in dire need of performance. Maybe sync with master and declare it ready for review?

ty for the ping on this one @Berndy , I will try to review #3417 in the next day or two

Harris-Miller commented:
@kedashoe made the updates you requested

kedashoe reviewed (approved):
:+1: Looks good to me, will leave open a bit for others to review the updates

Harris-Miller commented:
@kedashoe I added an optimization for arrays in `_clone`, but that broke unit tests so I reverted it. Will revisit that in another MR. Need re-approval because of the update

Harris-Miller commented:
@kedashoe all good again, read for re-approval
